adapter/memory/memory.go

```diff
@@ -0,0 +1,180 @@
+package memory
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/sankethkini/StudentData/domain/user"
+)
+
+//mapper for stroring existence of records
+//items for saving data in a datastructure
+type adapter struct {
+	mapper map[string]bool
+	Items  []user.User
+}
+
+//checking if a perticular record exists or not
+func (adpt *adapter) Retrive(field string, value string) bool {
+	return adpt.mapper[value]
+}
+
+//Search for right index
+func (adpt *adapter) searchForIndex(user user.User) int {
+	return sort.Search(len(adpt.Items), func(i int) bool {
+		if adpt.Items[i].Fname > user.Fname {
+			return true
+		} else if adpt.Items[i].Fname == user.Fname {
+			return adpt.Items[i].Fname < user.Fname
+		} else {
+			return false
+		}
+	})
+}
+
+//insert into map that record exists
+func (adpt *adapter) insertIntoMap(nums ...string) {
+
+	if adpt.mapper == nil {
+		adpt.mapper = make(map[string]bool)
+	}
+
+	for _, val := range nums {
+		adpt.mapper[val] = true
+	}
+}
+
+//insertUser for entry of single record
+func (adpt *adapter) insertUser(a interface{}) {
+
+	cur := a.(user.User)
+	if len(adpt.Items) == 0 {
+		adpt.Items = append(adpt.Items, cur)
+	} else {
+		i := adpt.searchForIndex(cur)
+		if i >= len(adpt.Items) {
+			adpt.Items = append(adpt.Items, cur)
+		} else {
+			adpt.Items = append(adpt.Items[:i+1], adpt.Items[i:]...)
+			adpt.Items[i] = cur
+		}
+
+	}
+	adpt.insertIntoMap(cur.RollNo)
+}
+
+//arraysorting all records entered once
+func (adpt *adapter) arraysort() {
+
+	sort.Slice(adpt.Items, func(i int, j int) bool {
+		if adpt.Items[i].Fname == adpt.Items[j].Fname {
+			return adpt.Items[i].RollNo < adpt.Items[j].RollNo
+		} else {
+			return adpt.Items[i].Fname < adpt.Items[j].Fname
+		}
+	})
+
+	var nums []string
+	for _, val := range adpt.Items {
+		nums = append(nums, val.RollNo)
+	}
+	adpt.insertIntoMap(nums...)
+}
+
+//save function
+//insertion sort for single entry
+//arraysorting if more than one record entry
+func (adpt *adapter) Save(item interface{}) error {
+
+	switch item := item.(type) {
+
+	case user.User:
+		adpt.insertUser(item)
+	case []user.User:
+		adpt.Items = append(adpt.Items, item...)
+		adpt.arraysort()
+	default:
+		return fmt.Errorf("error while saving records %w", NotARightType)
+	}
+
+	return nil
+}
+
+//reversing whole slice
+func reverse(arr []user.User) {
+	for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
+		arr[i], arr[j] = arr[j], arr[i]
+	}
+}
+
+//RetriveAll function to retrive all records
+//sorting based on field
+func (adpt *adapter) RetriveAll(field string, order int) (interface{}, error) {
```

Comment:
why does this func returns generic interface{} if its implemented only once and returns only single type of struct []User